rev_D functionality

raezt opened the pull request:
## Pull Request Description

Revision D firmware update.

## PR Type
- [ ] Bug fix (change that fixes an issue)
- [x] New feature (change that adds new functionality)
- [ ] Breaking change (has dependencies in other repos or will cause CI to fail)

## PR Checklist
- [x] I have followed the [Coding style guidelines](http://analogdevicesinc.github.io/no-OS/drivers_guide.html#coding-style)
- [x] I have performed a self-review of the changes
- [x] I have commented my code, at least hard-to-understand parts
- [x] I have build all projects affected by the changes in this PR
- [x] I have tested in hardware affected projects, at the relevant boards
- [x] I have signed off all commits from this PR
- [ ] I have updated the documentation (wiki pages, ReadMe etc), if applies

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/analogdevicesinc/no-OS?pullRequest=2215) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/analogdevicesinc/no-OS?pullRequest=2215) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: danmois<br/>:x: raezt<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/analogdevicesinc/no-OS?pullRequest=2215) it.</sub>

amiclaus reviewed (commented):
there are some build errors that need handling in the first place, prior to the review.
Also, what about REV_B, REV_C? Are those equivalent from the software point of view with REV_D? I'm asking because if that's the case, the software handling for the revisions might be simplified.

amiclaus commented on projects/ad-acevsecrdset-sl/src/inter/inter.c:
```diff
@@ -0,0 +1,375 @@
+	}
+
+	/* Initialize GPIO IRQ controller in order to be able to enable GPIO IRQ interrupt */
+	struct no_os_irq_init_param irq_gpio_ip = {
```
move this above together with the rest of the structure initializations.

amiclaus commented on projects/ad-acevsecrdset-sl/src/inter/inter.c:
```diff
@@ -0,0 +1,375 @@
+	ret = no_os_gpio_direction_input(pin);
+	if (ret) {
+		printf("ERROR\n");
+		return ret;
```
where is the memory deallocated for the `pin` in case of error? same applies for the rest of the gpio/irq descriptors.

amiclaus commented on projects/ad-acevsecrdset-sl/src/inter/inter.c:
```diff
@@ -0,0 +1,375 @@
+
+	ret = no_os_gpio_get(&pin2, &gpio_ip);
+	if (ret) {
+		printf("ERROR\n");
```
these code duplications can be optimized a bit with conditional jumps.

amiclaus commented on projects/ad-acevsecrdset-sl/src/self_test/self_test.c:
```diff
@@ -441,6 +470,26 @@ static int self_test_pilot(struct stout *stout)
+	} else
+		return INTF_PE_UPSTREAM_ERR;
+
+	return 0;
```
is this ever reached?

amiclaus commented on projects/ad-acevsecrdset-sl/src/self_test/self_test.c:
```diff
@@ -441,6 +470,26 @@ static int self_test_pilot(struct stout *stout)
+ * @param stout - state  machine descriptor
+ * @return 0 in case of success, error code otherwise
+ */
+int self_test_PE_upstream(struct stout *stout)
```
not sure about the previous implementation of the project, but snake case convention is applied.

amiclaus reviewed (commented):
some comments on my side.

raezt commented:
v3
- renamed self_test_PE_upstream
- modified the self_test_PE_upstream return
- moved "struct no_os_irq_init_param irq_gpio_ip" where the other structures are initialized

raezt commented:
v4

- squash 2 commits (the second one was fixing the previous one)

amiclaus commented on projects/ad-acevsecrdset-sl/Makefile:
```diff
@@ -1,7 +1,15 @@
+ifdef REV_A
+CFLAGS += -DREV_A=1
+else
+CFLAGS += -DREV_D=1
```
If there will be other Revisions in the future then we might want to keep both flags (REV_A and REV_D). Also drop the explicit initialization with 1 since it is redundant. To keep the approach similar to the rest of the projects, we should do something like this:
```
REV_A = n
REV_D = y
ifeq (y,$(strip $(REV_A)))
CFLAGS += -DREV_A
else
CFLAGS += -DREV_D
endif
```

raezt commented:
v5

- optimized the code with conditional jumps (interupt.c).
- removed the statement guards where there is only one expression.
- changed printf to pr_err.
- dropped the extra comment blocks.
- memory deallocated in case of error.

raezt commented:
v6

- squash 2 commits (the first one was implementing just a part of the added functionality)

raezt commented:
v7

- changed the makefile for revision

amiclaus reviewed (commented):
some more minor comments on my side. otherwise looks good.

raezt commented:
v8

- REV_A=true changed to REV_A=y in commit message.
- deleted line from eof in makefile.
- add Functions Definition header comment in inter.c.
- unrelated change in self_test.c resolved.
